### PR TITLE
feat(minor-service-registry)!: verifier details query

### PR DESCRIPTION
**Pull request by maancham**
## Description

AXE-4686
## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

### CODE REVIEW

**Review by cgorenflo — CHANGES_REQUESTED** [review]

File: contracts/service-registry/src/state.rs [hunk 1]
```diff
@@ -145,12 +145,48 @@ pub enum AuthorizationState {
...
+    (ServiceName, ChainName, VerifierAddress),
+    VerifierChainRecord,
+    VerifierPerChainIndexes,
+> = IndexedMap::new(
+    "verifiers",
+    VerifierPerChainIndexes {
+        verifier: MultiIndex::new(
+            |_pk: &[u8], d| d.verifier.clone(),
+            "verifiers",
+            "verifiers_service_verifier",
```
Comment (cgorenflo) [inline contracts/service-registry/src/state.rs:181]:
```suggestion
            "verifiers__address",
```

File: contracts/service-registry/src/state.rs [hunk 2]
```diff
@@ -145,12 +145,48 @@ pub enum AuthorizationState {
...
+
+#[cw_serde]
+pub struct VerifierChainRecord {
+    pub chain: ChainName,
+    pub verifier: VerifierAddress,
+}
+
+#[index_list(VerifierChainRecord)]
+pub struct VerifierPerChainIndexes<'a> {
+    pub verifier: MultiIndex<
```
Comment (cgorenflo) [inline contracts/service-registry/src/state.rs:163]:
```suggestion
    pub verifier_address: MultiIndex<
```

File: contracts/service-registry/src/state.rs (hunk 1 above)
Comment (cgorenflo) [reply contracts/service-registry/src/state.rs:181]: keys can have " _ " in them, so a single " _ " between pk namespace and index namespace might not be safe

**Review by cgorenflo — APPROVED** [review]

**Review by cgorenflo — CHANGES_REQUESTED** [review]
mistakenly clicked approve

File: contracts/service-registry/src/state.rs [hunk 3]
```diff
@@ -173,7 +211,17 @@ pub fn deregister_chains_support(
...
+            .remove(
+                storage,
+                (service_name.clone(), chain.clone(), verifier.clone()),
+            )
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Failed to deregister support for chain {:?} from verifier {:?}",
+                    chain, verifier
+                )
+            });
```
Comment (cjcobb23) [inline contracts/service-registry/src/state.rs:224]: I wouldn't panic here, just return the error. Use change_context if need be

File: contracts/service-registry/src/state.rs [hunk 4]
```diff
@@ -145,12 +145,46 @@ pub enum AuthorizationState {
...
+    "verifiers",
+    VerifierPerChainIndexes {
+        verifier_address: MultiIndex::new(
+            |pk: &[u8], _: &()| {
+                let (service_name, _, verifier) =
+                    <(ServiceName, ChainName, VerifierAddress)>::from_slice(pk)
+                        .expect("Invalid primary key");
+                (service_name, verifier)
+            },
+            "verifiers",
```
Comment (cjcobb23) [inline contracts/service-registry/src/state.rs:178]:
```suggestion
            "verifiers_per_chain",
```

File: contracts/service-registry/src/state.rs [hunk 5]
```diff
@@ -145,12 +145,46 @@ pub enum AuthorizationState {
...
+}
+
+impl<'a> IndexList<()> for VerifierPerChainIndexes<'a> {
+    fn get_indexes(&'_ self) -> Box<dyn Iterator<Item = &'_ dyn Index<()>> + '_> {
+        let v: Vec<&dyn Index<()>> = vec![&self.verifier_address];
+        Box::new(v.into_iter())
+    }
+}
+
+pub const VERIFIERS_PER_CHAIN_INDEXED_MAP: IndexedMap<
```
Comment (cjcobb23) [inline contracts/service-registry/src/state.rs:164]:
```suggestion
pub const VERIFIERS_PER_CHAIN: IndexedMap<
```

I generally don't like including the type in the name. Anyone can just go to the definition to see the type, or just use their IDE, so it feels unnecessarily verbose.

File: contracts/service-registry/src/contract/query.rs [hunk 6]
```diff
@@ -12,27 +23,35 @@ pub fn active_verifiers(
...
+            VERIFIERS
+                .may_load(deps.storage, (&service_name, &verifier_addr))
+                .ok()
+                .flatten()
+        })
+        .filter(|verifier| {
+            matches!(
+                verifier.bonding_state,
+                BondingState::Bonded { amount } if amount >= service.min_verifier_bond
+            ) && verifier.authorization_state == AuthorizationState::Authorized
```
Comment (cjcobb23) [inline contracts/service-registry/src/contract/query.rs:43]: why did any of this need to change?

File: contracts/service-registry/src/msg.rs [hunk 7]
```diff
@@ -83,6 +83,11 @@ pub enum QueryMsg {
         service_name: String,
         verifier: String,
     },
+    #[returns(crate::contract::query::VerifierDetailsResponse)]
+    VerifierDetails {
+        service_name: String,
+        verifier: String,
+    },
```
Comment (cjcobb23) [inline contracts/service-registry/src/msg.rs:90]: i wouldn't add a new query. Just change the return type of the `Verifier` query

Comment (cjcobb23) [reply contracts/service-registry/src/msg.rs:90]: also, imo it's clearer to define the return type here in the msg.rs file. Makes things easier to read.

**Review by cjcobb23 — CHANGES_REQUESTED** [review]
overall looks good, just a couple of things

File: contracts/service-registry/src/state.rs (hunk 3 above)
Comment (maancham) [reply contracts/service-registry/src/state.rs:224]: I used `map_err`.

File: contracts/service-registry/src/contract/query.rs [hunk 8]
```diff
@@ -36,21 +44,47 @@ pub fn active_verifiers(
...
+    let verifier_addr = deps.api.addr_validate(&verifier)?;
+
+    let verifier = VERIFIERS
+        .may_load(deps.storage, (&service_name, &verifier_addr))?
+        .ok_or(ContractError::VerifierNotFound)?;
+
+    let supported_chains = VERIFIERS_PER_CHAIN
+        .idx
+        .verifier_address
+        .prefix((service_name.clone(), verifier_addr.clone()))
```
Comment (cgorenflo) [inline contracts/service-registry/src/contract/query.rs:78]: do you need to clone the service name?

File: contracts/service-registry/src/contract/query.rs [hunk 9]
```diff
@@ -36,21 +44,47 @@ pub fn active_verifiers(
...
+        .keys(deps.storage, None, None, Order::Ascending)
+        .map(|result| result.map(|(_, chain, _)| chain))
+        .collect::<Result<Vec<ChainName>, _>>()?;
```
Comment (cgorenflo) [inline contracts/service-registry/src/contract/query.rs:81]: try_collect

File: contracts/service-registry/src/contract/query.rs [hunk 10]
```diff
@@ -36,21 +44,47 @@ pub fn active_verifiers(
...
+    let verifier = VERIFIERS
+        .may_load(deps.storage, (&service_name, &verifier_addr))?
+        .ok_or(ContractError::VerifierNotFound)?;
+
+    let supported_chains = VERIFIERS_PER_CHAIN
+        .idx
+        .verifier_address
+        .prefix((service_name.clone(), verifier_addr.clone()))
+        .keys(deps.storage, None, None, Order::Ascending)
+        .map(|result| result.map(|(_, chain, _)| chain))
```
Comment (cgorenflo) [inline contracts/service-registry/src/contract/query.rs:80]:
```suggestion
        .map_ok(|(_, chain, _)| chain)
```

File: contracts/service-registry/src/error.rs [hunk 11]
```diff
@@ -33,4 +33,6 @@ pub enum ContractError {
     NotEnoughVerifiers,
     #[error("verifier is jailed")]
     VerifierJailed,
+    #[error("Failed to deregister chain support")]
```
Comment (cgorenflo) [inline contracts/service-registry/src/error.rs:36]:
```suggestion
    #[error("failed to deregister chain support")]
```

File: contracts/service-registry/src/msg.rs [hunk 12]
```diff
@@ -66,6 +71,13 @@ pub enum ExecuteMsg {
     ClaimStake { service_name: String },
 }
 
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct VerifierDetailsResponse {
+    pub verifier: Verifier,
+    pub weight: nonempty::Uint128,
+    pub supported_chains: Vec<ChainName>,
+}
+
```
Comment (cgorenflo) [inline contracts/service-registry/src/msg.rs:80]: move after the QueryMsg

File: contracts/service-registry/src/state.rs [hunk 13]
```diff
@@ -145,12 +145,46 @@ pub enum AuthorizationState {
...
+        (ServiceName, ChainName, VerifierAddress),
+    >,
+}
+
+impl<'a> IndexList<()> for VerifierPerChainIndexes<'a> {
+    fn get_indexes(&'_ self) -> Box<dyn Iterator<Item = &'_ dyn Index<()>> + '_> {
+        let v: Vec<&dyn Index<()>> = vec![&self.verifier_address];
+        Box::new(v.into_iter())
+    }
+}
```
Comment (cgorenflo) [inline contracts/service-registry/src/state.rs:162]: didn't you use a macro before for this?

File: contracts/service-registry/src/state.rs [hunk 14]
```diff
@@ -145,12 +145,46 @@ pub enum AuthorizationState {
...
+    VerifierPerChainIndexes {
+        verifier_address: MultiIndex::new(
+            |pk: &[u8], _: &()| {
+                let (service_name, _, verifier) =
+                    <(ServiceName, ChainName, VerifierAddress)>::from_slice(pk)
+                        .expect("Invalid primary key");
+                (service_name, verifier)
+            },
+            "verifiers_per_chain",
+            "verifiers__address",
```
Comment (cgorenflo) [inline contracts/service-registry/src/state.rs:179]:
```suggestion
            "verifiers_per_chain__address",
```

File: contracts/service-registry/src/state.rs [hunk 15]
```diff
@@ -173,9 +210,10 @@ pub fn deregister_chains_support(
     verifier: VerifierAddress,
 ) -> Result<(), ContractError> {
     for chain in chains {
-        VERIFIERS_PER_CHAIN.remove(storage, (&service_name, &chain, &verifier));
+        VERIFIERS_PER_CHAIN
+            .remove(storage, (service_name.clone(), chain, verifier.clone()))
+            .map_err(|_| ContractError::DeregistrationFailed)?;
```
Comment (cgorenflo) [inline contracts/service-registry/src/state.rs:215]: why change the error here? None of the other function have that

File: contracts/service-registry/src/state.rs [hunk 16]
```diff
@@ -145,12 +145,46 @@ pub enum AuthorizationState {
...
+    (),
+    VerifierPerChainIndexes,
+> = IndexedMap::new(
+    "verifiers_per_chain",
+    VerifierPerChainIndexes {
+        verifier_address: MultiIndex::new(
+            |pk: &[u8], _: &()| {
+                let (service_name, _, verifier) =
+                    <(ServiceName, ChainName, VerifierAddress)>::from_slice(pk)
+                        .expect("Invalid primary key");
```
Comment (cgorenflo) [inline contracts/service-registry/src/state.rs:175]:
```suggestion
                        .expect("invalid primary key");
```

**Review by cgorenflo — CHANGES_REQUESTED** [review]

File: contracts/service-registry/src/msg.rs [hunk 17]
```diff
@@ -66,6 +71,13 @@ pub enum ExecuteMsg {
     ClaimStake { service_name: String },
 }
 
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct VerifierDetailsResponse {
```
Comment (cjcobb23) [inline contracts/service-registry/src/msg.rs:75]:
```suggestion
pub struct VerifierDetails {
```
up to you, but I feel like this is cleaner. The `Response` suffix always feels so ugly to me

File: contracts/service-registry/src/contract/query.rs [hunk 18]
```diff
@@ -13,19 +15,25 @@ pub fn active_verifiers(
...
+            VERIFIERS
+                .may_load(deps.storage, (&service_name, &verifier_addr))
+                .ok()
+                .flatten()
+        })
+        .filter(|verifier| {
+            matches!(
+                verifier.bonding_state,
+                BondingState::Bonded { amount } if amount >= service.min_verifier_bond
+            ) && verifier.authorization_state == AuthorizationState::Authorized
```
[the remark this answers is not in the document]
Comment (cjcobb23) [reply contracts/service-registry/src/contract/query.rs:32]: yes I am still confused why this function needed to change so much

File: contracts/service-registry/src/contract/query.rs [hunk 19]
```diff
@@ -13,19 +15,25 @@ pub fn active_verifiers(
...
         })
-        .filter(|verifier| verifier.authorization_state == AuthorizationState::Authorized)
         .map(|verifier| WeightedVerifier {
             verifier_info: verifier,
-            weight: VERIFIER_WEIGHT, // all verifiers have an identical const weight for now
```
Comment (cjcobb23) [inline contracts/service-registry/src/contract/query.rs:28]: i think we should keep this comment

File: contracts/service-registry/src/contract.rs [hunk 20]
```diff
@@ -2,18 +2,18 @@ use axelar_wasm_std::{address, permission_control, FnExt};
...
 
 use crate::error::ContractError;
 use crate::msg::{ExecuteMsg, InstantiateMsg, QueryMsg};
-use crate::state::{AuthorizationState, BondingState, Service, Verifier, SERVICES, VERIFIERS};
+use crate::state::{AuthorizationState, BondingState, Service, SERVICES, VERIFIERS};
 
 mod execute;
 mod migrations;
-mod query;
+pub mod query;
```
[the remark this answers is not in the document]
Comment (maancham) [reply contracts/service-registry/src/contract.rs:16]: the returned schema needs to be imported on coordinator (later)?

File: contracts/service-registry/src/contract/query.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (maancham) [reply contracts/service-registry/src/contract/query.rs:32]: the previous schema returned verifier object from the map, but with our indexed map, we can only have the address, so I need to first grab the actual verifier.

File: contracts/service-registry/src/state.rs (hunk 13 above)
Comment (maancham) [reply contracts/service-registry/src/state.rs:162]: apparently, that macro does not work when applied to empty value (the updated version of indexed map does not have a value). When added with empty value, build fails saying we need to implement IndexList

File: contracts/service-registry/src/state.rs (hunk 15 above)
Comment (maancham) [reply contracts/service-registry/src/state.rs:215]: removed.

File: contracts/service-registry/src/msg.rs (hunk 17 above)
Comment (maancham) [reply contracts/service-registry/src/msg.rs:75]: yeah makes sense. Updated!

File: contracts/service-registry/src/state.rs [hunk 21]
```diff
@@ -143,12 +143,46 @@ pub enum AuthorizationState {
...
+
+impl<'a> IndexList<()> for VerifierPerChainIndexes<'a> {
+    fn get_indexes(&'_ self) -> Box<dyn Iterator<Item = &'_ dyn Index<()>> + '_> {
+        let v: Vec<&dyn Index<()>> = vec![&self.verifier_address];
+        Box::new(v.into_iter())
+    }
+}
+
+pub const VERIFIERS_PER_CHAIN: IndexedMap<
+    (ServiceName, ChainName, VerifierAddress),
```
Comment (cgorenflo) [inline contracts/service-registry/src/state.rs:163]: is it not possible to use references for keys anymore?

File: contracts/service-registry/src/contract/query.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (maancham) [reply contracts/service-registry/src/contract/query.rs:33]: Thanks for the suggestion!

File: contracts/service-registry/src/state.rs (hunk 21 above)
Comment (maancham) [reply contracts/service-registry/src/state.rs:163]: Yeah it is! Changed to references.

**Review by cgorenflo — APPROVED** [review]

**Review by cjcobb23 — APPROVED** [review]

**Pull request merged**